include::../description.adoc[]

== Noncompliant Code Example

https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.aesmanaged?view=netframework-4.8[AesManaged] object with insecure mode:

[source,vbnet]
----
Dim aes = New AesManaged With {
    .KeySize = 256,
    .BlockSize = 256,
    .Mode = CipherMode.ECB,
    .Padding = PaddingMode.PKCS7
} ' Noncompliant
----

https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.rsacryptoserviceprovider?view=netframework-4.8[RSACryptoServiceProvider] object without OAEP padding:

[source,vbnet]
----
Dim RSA1 = New RSACryptoServiceProvider()
Dim encryptedData = RSA1.Encrypt(dataToEncrypt, False) ' Noncompliant: OAEP Padding is not used (second parameter set to false)
----

== Compliant Solution

AES with GCM mode with https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.aesgcm?view=netcore-3.0[AesGcm] object: 

[source,vbnet]
----
Dim aesGcm = New AesGcm(key)
----

AES with GCM mode with https://www.bouncycastle.org/[bouncycastle] library: 

[source,vbnet]
----
Dim blockCipher As GcmBlockCipher = New GcmBlockCipher(New AesFastEngine())
blockCipher.Init(True, New AeadParameters(New KeyParameter(secretKey), 256, iv, Nothing))
----

If AES with GCM mode is not available, use AES CBC without padding or chose an alternative cipher mode: 

[source,vbnet]
----
Dim aesCBC = New AesManaged With {
    .KeySize = 256,
    .BlockSize = 256,
    .Mode = CipherMode.CBC,
    .Padding = PaddingMode.None
}

Dim aesCBC = New AesManaged With {
    .KeySize = 256,
    .BlockSize = 256,
    .Mode = CipherMode.CTS
}
----


RSA with OAEP padding with https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.rsacryptoserviceprovider?view=netframework-4.8[RSACryptoServiceProvider] object: 

[source,vbnet]
----
Dim RSA = New RSACryptoServiceProvider()
Dim encryptedData = RSA.Encrypt(dataToEncrypt, True)
----

include::../see.adoc[]

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

include::message.adoc[]

include::highlighting.adoc[]

'''
== Comments And Links
(visible only on this page)

include::../comments-and-links.adoc[]
endif::env-github,rspecator-view[]
